tsdb/head_test.go

```diff
@@ -727,49 +727,111 @@ func TestHead_Truncate(t *testing.T) {
+		truncateBefore int
+	}
 
-	memChunkPool := sync.Pool{
-		New: func() interface{} {
-			return &memChunk{}
+	testCases := []testCaseT{
+		{
+			chunkRange:     2000,
+			mmappedSamples: 3000,
+			headSamples:    0,
+			truncateBefore: 2000,
+		},
+		{
+			chunkRange:     2000,
+			mmappedSamples: 4000,
+			headSamples:    0,
+			truncateBefore: 2000,
+		},
+		{
+			chunkRange:     2000,
+			mmappedSamples: 4000,
+			headSamples:    6000,
+			truncateBefore: 2000,
+		},
+		{
+			chunkRange:     2000,
+			mmappedSamples: 4000,
+			headSamples:    6000,
+			truncateBefore: 4000,
+		},
+		{
+			chunkRange:     2000,
+			mmappedSamples: 4000,
+			headSamples:    10000,
+			truncateBefore: 8000,
 		},
 	}
 
-	s := newMemSeries(labels.FromStrings("a", "b"), 1, defaultIsolationDisabled)
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("range=%d mmappedSamples=%d headSamples=%d truncateBefore=%d", tc.chunkRange, tc.mmappedSamples, tc.headSamples, tc.truncateBefore), func(t *testing.T) {
+			dir := t.TempDir()
+			// This is usually taken from the Head, but passing manually here.
+			chunkDiskMapper, err := chunks.NewChunkDiskMapper(nil, dir, chunkenc.NewPool(), chunks.DefaultWriteBufferSize, chunks.DefaultWriteQueueSize)
+			require.NoError(t, err)
+			defer func() {
+				require.NoError(t, chunkDiskMapper.Close())
+			}()
 
-	for i := 0; i < 4000; i += 5 {
-		ok, _ := s.append(int64(i), float64(i), 0, chunkDiskMapper, chunkRange)
-		require.True(t, ok, "sample append failed")
-	}
+			memChunkPool := sync.Pool{
+				New: func() interface{} {
+					return &memChunk{}
+				},
+			}
 
-	// Check that truncate removes half of the chunks and afterwards
-	// that the ID of the last chunk still gives us the same chunk afterwards.
-	countBefore := len(s.mmappedChunks) + 1 // +1 for the head chunk.
-	lastID := s.headChunkID(countBefore - 1)
-	lastChunk, _, err := s.chunk(lastID, chunkDiskMapper, &memChunkPool)
-	require.NoError(t, err)
-	require.NotNil(t, lastChunk)
+			s := newMemSeries(labels.FromStrings("a", "b"), 1, defaultIsolationDisabled)
 
-	chk, _, err := s.chunk(0, chunkDiskMapper, &memChunkPool)
-	require.NotNil(t, chk)
-	require.NoError(t, err)
+			// append samples that will end up in mmapped chunks
+			if tc.mmappedSamples > 0 {
+				for i := 0; i < tc.mmappedSamples; i += 5 {
+					ok, _ := s.append(int64(i), float64(i), 0, chunkDiskMapper, int64(tc.chunkRange))
+					require.True(t, ok, "sample append failed")
+				}
+				s.mmapHeadChunks(chunkDiskMapper)
+				require.Greater(t, len(s.mmappedChunks), 0)
+			}
 
-	s.truncateChunksBefore(2000, 0)
+			// append samples that will stay in the head
+			if tc.headSamples > 0 {
+				for i := tc.mmappedSamples; i < tc.mmappedSamples+tc.headSamples; i += 5 {
+					ok, _ := s.append(int64(i), float64(i), 0, chunkDiskMapper, int64(tc.chunkRange))
+					require.True(t, ok, "sample append failed")
+				}
+				require.Greater(t, len(s.headChunks), 0)
+			}
```

Comment:
You can simplify this by having a loop from 0 to tc.headSamples with a condition inside like
```
if i == tc.mmappedSamples {
    // mmap
    // check for >0
} 
```

(while you do that, can you change the comment to start with a capital letter where possible and end with a `.`?)